src/sql/parts/notebook/models/notebookModel.ts

```diff
@@ -469,20 +539,26 @@ export class NotebookModel extends Disposable implements INotebookModel {
 	}
 
 	public changeKernel(displayName: string): void {
-		let spec = this.getKernelSpecFromDisplayName(displayName);
-		this.doChangeKernel(spec);
+		this._contextsLoadingEmitter.fire();
+		this.doChangeKernel(displayName, true);
 	}
 
-	public doChangeKernel(kernelSpec: nb.IKernelSpec): Promise<void> {
-		this.setProviderIdForKernel(kernelSpec);
-		// Ensure that the kernel we try to switch to is a valid kernel; if not, use the default
-		let kernelSpecs = this.getKernelSpecs();
-		if (kernelSpecs && kernelSpecs.length > 0 && kernelSpecs.findIndex(k => k.name === kernelSpec.name) < 0) {
-			kernelSpec = kernelSpecs.find(spec => spec.name === this.notebookManager.sessionManager.specs.defaultKernel);
+	public async doChangeKernel(displayName: string, needSetProvider: boolean = true): Promise<void> {
```

Comment:
minor: `mustSetProvider` instead of `needSet`

Reply:
good;)